### PR TITLE
i#5984: Fix trace invariant false positive on back-to-back signals

### PR DESCRIPTION
**Pull request by abhinav92003**
Moves tracking for previous kernel_event marker values and previous pre-signal instrs at the kernel_event and kernel_xfer markers. Previously we did some part of this book-keeping at instrs after the kernel_xfer marker, which would be erroneously skipped when there are back-to-back signals without any intervening instrs.

Adds tracking for last seen instr at each signal depth. This is required to correctly find the pre-signal instr for the second signal in a back-to-back signal pair. Without this, the pre-signal instr for the second signal is erroneously assumed to be whatever the last instr is, which may be some instr in the first signal's body. This also obviates the need to track the boolean for saw_kernel_xfer_after_prev_instr_, which was to detect nested signals Without any intervening instr.

Adds a test case that verifies the scenario with back-to-back signals without any intervening instr.

Fixes: #5984

### CODE REVIEW

**Comment by abhinav92003** [comment]
AArch64 failure is #5635.

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 1]
```diff
@@ -328,6 +331,73 @@ check_kernel_xfer()
...
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Trace starts in a signal.
+    {
+        std::vector<memref_t> memrefs = {
+            // Already inside the first signal.
+            gen_instr(1, 11),
+            // XXX: This marker value is actually not guaranteed, yet the checker
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:367]: What does "not guaranteed" mean?  Why is this different from any other sigreturn/etc.?  This marker value is required to be correct and the checker should be checking it.

Ditto below.

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 2]
```diff
@@ -328,6 +331,73 @@ check_kernel_xfer()
...
+            // requires it and the view tool prints it.
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_XFER, 12),
+            // No intervening instr here. Should skip pre-signal instr check on
+            // return; this is a special case as it would require *removing* the
+            // pc = 1 instr from pre_signal_instr_ as it was not in this newly
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:389]: ?  There is no pc=1 instr here: you mean pc=11?

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 3]
```diff
@@ -370,6 +440,8 @@ check_rseq()
             gen_marker(1, TRACE_MARKER_TYPE_RSEQ_ABORT, 2),
             gen_marker(1, TRACE_MARKER_TYPE_KERNEL_EVENT, 2),
             gen_marker(1, TRACE_MARKER_TYPE_KERNEL_EVENT, 4),
+            gen_instr(1, 10),
```
Comment (derekbruening) [inline clients/drcachesim/tests/invariant_checker_test.cpp:443]: Does the checker fail without these additions?  What if the trace did actually end before these were added?

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 4]
```diff
@@ -439,25 +454,36 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
+                     memref.instr.addr == shard->last_xfer_int_pc_ ||
+                     // DR hands us a different address for sysenter than the
+                     // resumption point.
+                     shard->last_pre_signal_instr_.instr.type ==
+                         TRACE_TYPE_INSTR_SYSENTER) &&
                  (
-                     // Skip pre_signal_instr_ check if there was no such instr.
-                     shard->pre_signal_instr_.top().instr.addr == 0 ||
-                     memref.instr.addr == shard->pre_signal_instr_.top().instr.addr ||
+                     // Skip pre-signal instr check if there was no such instr.
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:472]: Should we split these into two different error strings: the post==marker (above) with one error string and the pre==post (below) with a different one?  May make it easier to analyze/group reported invariants, and split the complex boolean expression here up too.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 5]
```diff
@@ -439,25 +454,36 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
-                  // DR hands us a different address for sysenter than the
-                  // resumption point.
-                  shard->pre_signal_instr_.top().instr.type ==
-                      TRACE_TYPE_INSTR_SYSENTER) &&
+                ((
+                     // Skip this check if we did not see the corresponding
+                     // kernel_event marker in the trace because the trace
+                     // started mid-signal.
+                     shard->last_xfer_int_pc_ == 0 ||
+                     memref.instr.addr == shard->last_xfer_int_pc_ ||
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:466]: With the large comments on the other conditions around this, I found it hard to identify this as its own thing rather than part of the condition above.  Maybe put this first, or add a comment saying this is the regular check for post==marker.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 6]
```diff
@@ -439,25 +454,36 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
+                     // handler, verified using last_xfer_int_pc_ above.
+                     shard->last_xfer_aborted_rseq_ ||
+                     memref.instr.addr == shard->last_pre_signal_instr_.instr.addr ||
                      // Asynch will go to the subsequent instr.
                      memref.instr.addr ==
-                         shard->pre_signal_instr_.top().instr.addr +
-                             shard->pre_signal_instr_.top().instr.size ||
+                         shard->last_pre_signal_instr_.instr.addr +
+                             shard->last_pre_signal_instr_.instr.size ||
                      // Too hard to figure out branch targets.  We have the
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:483]: But we do have branch targets now: it's right there in a shard var, isn't it?

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 7]
```diff
@@ -469,17 +495,18 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
-            // We assume paired signal entry-exit (so no longjmp and no rseq
-            // inside signal handlers).
-            shard->prev_xfer_int_pc_.pop();
-            shard->pre_signal_instr_.pop();
         }
+        // It is a little inefficient to replace the top instr on the stack everytime.
+        // But we cannot perform this book-keeping using prev_instr_ on a
+        // TRACE_MARKER_TYPE_KERNEL_EVENT marker. If there was no instruction between
+        // two nested signals, we do not want to record any pre-signal instr for the
+        // second signal.
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:503]: "between two nested signals": there is no test of this added here so I wonder if you mean "between two consecutive signals at the same scope" (i.e., not nested wrt each other).

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 8]
```diff
@@ -519,26 +550,52 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
+                // forget the last instr we saw in what was incorrectly assumed previously
+                // to be the outer-most context.
+                replace_top<memref_t>(shard->pre_signal_instr_, {});
+            } else {
+                shard->last_xfer_int_pc_ = shard->prev_xfer_int_pc_.top();
+                shard->prev_xfer_int_pc_.pop();
+                shard->pre_signal_instr_.pop();
+                // The last instr in the context immediately outer to this signal may
+                // be {} in some cases:
+                // - for nested signals without any intervening instr
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:580]: See comment in the test: there is no test of this.  Do you mean back-to-back signals at the same scope instead of nested signals?

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 9]
```diff
@@ -469,17 +495,18 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
                 "Signal handler return point incorrect");
-            // We assume paired signal entry-exit (so no longjmp and no rseq
-            // inside signal handlers).
-            shard->prev_xfer_int_pc_.pop();
-            shard->pre_signal_instr_.pop();
         }
+        // It is a little inefficient to replace the top instr on the stack everytime.
+        // But we cannot perform this book-keeping using prev_instr_ on a
+        // TRACE_MARKER_TYPE_KERNEL_EVENT marker. If there was no instruction between
+        // two nested signals, we do not want to record any pre-signal instr for the
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:502]: But we do want the pre-signal instr for an instr prior to two consecutive signals at the same scope.  What we want is a prev_instr stack so we get the prev_instr for our scope -- but isn't that what shard->pre_signal_instr is already?  I think I'm confused and not understanding this right: not understanding this replace_top.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 10]
```diff
@@ -519,26 +550,52 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
+                shard->last_xfer_int_pc_ = shard->prev_xfer_int_pc_.top();
+                shard->prev_xfer_int_pc_.pop();
+                shard->pre_signal_instr_.pop();
+                // The last instr in the context immediately outer to this signal may
+                // be {} in some cases:
+                // - for nested signals without any intervening instr
+                // - if there's a signal at the very beginning of the trace
+                // In both these cases the empty instr implies that it should not
+                // be used for the pre-signal instr check.
+                shard->last_pre_signal_instr_ = shard->pre_signal_instr_.top();
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:584]: Why do some of these 3 stacks do "last=top; pop" while others do "pop; last=top"?  Shouldn't they all be the same wrt whether the top or second slot is what we want?

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 11]
```diff
@@ -519,26 +550,52 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
+                shard->last_xfer_aborted_rseq_ = shard->prev_xfer_aborted_rseq_.top();
+                shard->prev_xfer_aborted_rseq_.pop();
+            }
+        }
+        if (memref.marker.marker_type == TRACE_MARKER_TYPE_KERNEL_EVENT) {
+            // If preceded by an RSEQ abort marker, this is not really a signal.
+            if (shard->prev_entry_.marker.marker_type == TRACE_MARKER_TYPE_RSEQ_ABORT) {
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:591]: Doesn't this need to first check that prev_entry_.marker.type is a marker?

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 12]
```diff
@@ -553,6 +610,13 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
     }
 
 #ifdef UNIX
+    if (saw_rseq_abort) {
+        shard->saw_rseq_abort_ = true;
+    } else if (!(memref.marker.type == TRACE_TYPE_MARKER &&
+                 (memref.marker.marker_type == TRACE_MARKER_TYPE_TIMESTAMP ||
+                  memref.marker.marker_type == TRACE_MARKER_TYPE_CPU_ID))) {
```
Comment (derekbruening) [inline clients/drcachesim/tools/invariant_checker.cpp:617]: What is the logic for clearing on anything except a timestamp/cpuid?

**Review by derekbruening — COMMENTED** [review]
I think I'm confused about some aspects here.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 1 above)
Comment (abhinav92003) [reply clients/drcachesim/tests/invariant_checker_test.cpp:367]: This comment is same as many other similar comments in the test file. I think it points to the fact that a sigreturn is not necessary when coming back from a signal.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 2 above)
Comment (abhinav92003) [reply clients/drcachesim/tests/invariant_checker_test.cpp:389]: Ah yes. It should be pc = 11.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 3 above)
Comment (abhinav92003) [reply clients/drcachesim/tests/invariant_checker_test.cpp:443]: It does not fail.

As mentioned in the PR description, this is to verify that we don't raise a "signal handler return point incorrect" because of the apparent discontinuity between the post-signal instr and the instr just before the rseq-abort-causing signal.

See the paragraph with "Modified an existing invariant checker unit test to verify this scenario also." in the PR description.

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 13]
```diff
@@ -328,6 +331,73 @@ check_kernel_xfer()
         if (!run_checker(memrefs, false))
             return false;
     }
+    // Multiple nested signals without any intervening instr between the nested
+    // signals.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(1, 1),
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_EVENT, 2),
+            gen_instr(1, 101),
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/invariant_checker_test.cpp:340]: There's already a test for "Signal before any instr in the trace." and "Nested signals without any intervening instr". I'll add one with both in the same test.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 6 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:483]: Do you mean I should change it in this PR itself? Since this PR is already a bit complex, maybe we can do it in a follow up PR.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 14]
```diff
@@ -519,26 +550,52 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
-                // We have nested signals without an intervening app instr.
-                // Push an empty instr to mean that this shouldn't be used.
-                shard->pre_signal_instr_.push({});
+        if (memref.marker.marker_type == TRACE_MARKER_TYPE_KERNEL_XFER) {
+            // We assume paired signal entry-exit (so no longjmp and no rseq
+            // inside signal handlers).
+            if (shard->prev_xfer_int_pc_.empty() ||
+                shard->pre_signal_instr_.size() == 1 ||
+                shard->prev_xfer_aborted_rseq_.empty()) {
+                // If one is true, all should be so.
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:561]: If the trace starts with KERNEL_EVENT, all three of them will have something pushed. See L596.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 7 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:503]: We have test of both cases: see "Nested signals without any intervening instr." and "Multiple nested signals without any intervening instr between the nested".

> I wonder if you mean "between two consecutive signals at the same scope" (i.e., not nested wrt each other).

That's correct. I'll rephrase it to be clearer.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 8 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:580]: Here, I do mean nested signals. See "Nested signals without any intervening instr." in the test file.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 9 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:502]: > But we do want the pre-signal instr for an instr prior to two consecutive signals at the same scope

That's right. We do do that here. See comment in the "Multiple nested signals without any intervening instr between the nested" test.

> but isn't that what shard->pre_signal_instr is already?

Yes.

> not understanding this replace_top

When we see a new instr, that instr becomes the last instr seen in the current scope. That's what the replacement means.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 10 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:584]: last_pre_signal_instr_ needs to get the second-to-top (the instr in the scope that was interrupted by that signal).

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 12 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:617]: To detect whether the current signal caused an rseq abort we need to find the following pattern:
```
rseq_abort
kernel_event
kernel_event
```

I found that sometimes the two kernel events are separated by timestamp+cpu.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 11 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:591]: That's right. I'll add that.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tests/invariant_checker_test.cpp:367]: OK, so `s/This marker value/This marker/` and `s/guaranteed/guaranteed to be present/` -- though this is a particular sequence in a test so I would just remove this comment altogether.

Comment (abhinav92003) [reply clients/drcachesim/tests/invariant_checker_test.cpp]: Removed.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 2 above)
Comment (abhinav92003) [reply clients/drcachesim/tests/invariant_checker_test.cpp]: Done.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/invariant_checker_test.cpp:357]: Done.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 15]
```diff
@@ -36,6 +36,7 @@
 #include <algorithm>
 #include <iostream>
 #include <string.h>
+#include <cassert>
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:38]: Done.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 4 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp]: I split the boolean expression to simplify it.

But I think the different error messages are being worked on already in Sahil's PR. So I'll leave it to that.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 5 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp]: Added comment.

File: clients/drcachesim/tools/invariant_checker.cpp [hunk 16]
```diff
@@ -439,25 +454,36 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
...
-                     // Skip pre_signal_instr_ check if there was no such instr.
-                     shard->pre_signal_instr_.top().instr.addr == 0 ||
-                     memref.instr.addr == shard->pre_signal_instr_.top().instr.addr ||
+                     // Skip pre-signal instr check if there was no such instr.
+                     shard->last_pre_signal_instr_.instr.addr == 0 ||
+                     // Skip pre_signal_instr_ check for signals that caused an rseq
+                     // abort. In this case, control is transferred directly to the abort
+                     // handler, verified using last_xfer_int_pc_ above.
+                     shard->last_xfer_aborted_rseq_ ||
+                     memref.instr.addr == shard->last_pre_signal_instr_.instr.addr ||
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp]: Added the comment.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 7 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp]: Actually this comment applies to both: nested signals and consecutive signals. I clarified in the comment.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 11 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp]: Done.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 12 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp]: Added a comment to clarify.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 10 above)
Comment (derekbruening) [reply clients/drcachesim/tools/invariant_checker.cpp:588]: I don't follow this: what's at the top then?

I'm thinking we should have a single stack with a compound data structure, rather than a bunch of separate stacks which we have to worry about being out of sync.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tools/invariant_checker.cpp:565]: See below: a single stack with a compound data structure sounds like the right solution.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 9 above)
Comment (derekbruening) [reply clients/drcachesim/tools/invariant_checker.cpp:503]: What we want is shard->cur_scope->prev_instr.  Instead of popping and pushing to store prev_instr, if we do the compound data structure below we can more simply store into the cur scope (== top of scope/context stack)'s field for prev_instr.

Should we go further and get rid of shard->prev_instr in general and always use shard->cur_scope->prev_instr everywhere?

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 10 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:588]: > I don't follow this: what's at the top then?

The latest instruction in the current context (current signal).

> I'm thinking we should have a single stack with a compound data structure,

That does seem simpler but it may create some more complexities IMHO: we'd need to look at two entries of the stack for the operation here (since we want the second-to-top for pre_signal_instr_). Also updating the top instr (replace_top) will become a bit more expensive since it's a bigger entry.

> we have to worry about being out of sync.

We push/pop all at the same time always. 

If you feel strongly about this, I can try changing it. But I don't see a clear win in clarity.

**Review by derekbruening — COMMENTED** [review]
Should have done the last few comments as a review instead of separately.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 10 above)
Comment (derekbruening) [reply clients/drcachesim/tools/invariant_checker.cpp:588]: > The latest instruction in the current context (current signal).

But isn't that the current record?  So this is just an ordering problem where it's already set the prev to equal the current?

Comment (derekbruening) [reply clients/drcachesim/tools/invariant_checker.cpp:588]: > We push/pop all at the same time always.

This seems error-prone.

> If you feel strongly about this, I can try changing it. But I don't see a clear win in clarity.

Yes I do think a single stack is a better solution.  The code is simpler and clearer; the invariants are clearer (no more multiple comments and asserts about how all these separate stacks are supposed to be coordinated); changing an aspect of the top of the stack looks simpler (set a field, instead of pop;push).

Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:588]: Thinking more on this:  I think this can be simplified further. I'll add something like `last_instr_in_cur_context` which can repeatedly be updated, and then keep all context for the current signal in the same entry on one stack.

Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp]: Done.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 9 above)
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:493]: I switched to using a separate var `last_instr_in_cur_context_` that keeps getting updated.

Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp:493]: > Should we go further and get rid of shard->prev_instr in general and always use shard->cur_scope->prev_instr everywhere?

That'd get a little complex. `shard->cur_scope->prev_instr ` (or shard->signal_stack_.top().pre_signal_instr in the code now) may not be set. It can be `{ }` (to denote absence of any pre-signal instr in that context). So we'll need to search the whole signal_stack_ from the top to find the first entry with a valid pre_signal_instr.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 8 above)
Comment (derekbruening) [reply clients/drcachesim/tools/invariant_checker.cpp:562]: Right, I think I was confused before, this makes sense.

**Review by derekbruening — APPROVED** [review]
Thanks for making the change to a single stack: IMHO it's nicer now.  LGTM.

File: clients/drcachesim/tools/invariant_checker.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tools/invariant_checker.cpp]: Done.

**Comment by abhinav92003** [comment]
> Thanks for making the change to a single stack: IMHO it's nicer now. LGTM.

Yes, thanks for the suggestion. Combined with the separate last_instr_in_cur_context, it does become much simpler
